vectorstores/alloydb/vectorestores.go

```diff
@@ -0,0 +1,246 @@
+	}
+	embedderData, err := embedder.EmbedQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	whereQuerys := make([]string, 0)
+	if scoreThreshold != 0 {
+		whereQuerys = append(whereQuerys, fmt.Sprintf("data.distance < %f", 1-scoreThreshold))
+	}
+	for k, v := range filter {
+		whereQuerys = append(whereQuerys, fmt.Sprintf("(data.cmetadata ->> '%s') = '%s'", k, v))
+	}
+	whereQuery := strings.Join(whereQuerys, " AND ")
+	if len(whereQuery) == 0 {
+		whereQuery = "TRUE"
+	}
+	dims := len(embedderData)
+	sql := fmt.Sprintf(`WITH filtered_embedding_dims AS MATERIALIZED (
+    SELECT
+        *
+    FROM
+        %s
+    WHERE
+        vector_dims (
+                embedding
+        ) = $1
+)
+SELECT
+	data.document,
+	data.cmetadata,
+	(1 - data.distance) AS score
+FROM (
+	SELECT
+		filtered_embedding_dims.*,
+		embedding <=> $2 AS distance
+	FROM
+		filtered_embedding_dims
+		JOIN %s ON filtered_embedding_dims.collection_id=%s.uuid WHERE %s.name='%s') AS data
+WHERE %s
+ORDER BY
+	data.distance
+LIMIT $3`, vs.embeddingTableName,
+		vs.collectionTableName, vs.collectionTableName, vs.collectionTableName, vs.collectionName,
+		whereQuery)
+
+	rows, err := vs.engine.pool.Query(ctx, sql, dims, pgvector.NewVector(embedderData), numDocuments)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	docs := make([]schema.Document, 0)
+	for rows.Next() {
+		doc := schema.Document{}
+		if err := rows.Scan(&doc.PageContent, &doc.Metadata, &doc.Score); err != nil {
+			return nil, err
+		}
+		docs = append(docs, doc)
+	}
+	return docs, rows.Err()
+}
+
+func getOptions(options ...vectorstores.Option) vectorstores.Options {
+	opts := vectorstores.Options{}
+	for _, opt := range options {
+		opt(&opts)
+	}
+	return opts
+}
+
+func validateScoreThreshold(scoreThreshold float32) (float32, error) {
+	if scoreThreshold < 0 || scoreThreshold > 1 {
+		return 0, errors.New("score threshold must be between 0 and 1")
+	}
+	return scoreThreshold, nil
+}
+
+// getFilters return metadata filters, now only support map[key]value pattern
+// TODO: should support more types like {"key1": {"key2":"values2"}} or {"key": ["value1", "values2"]}.
+func getFilters(opts vectorstores.Options) (map[string]any, error) {
+	if opts.Filters != nil {
+		if filters, ok := opts.Filters.(map[string]any); ok {
+			return filters, nil
+		}
+		return nil, errors.New("invalid filters")
+	}
+	return map[string]any{}, nil
+}
+
+func deduplicate(ctx context.Context, opts vectorstores.Options, docs []schema.Document) []schema.Document {
+	if opts.Deduplicater == nil {
+		return docs
+	}
+	filtered := make([]schema.Document, 0, len(docs))
+	for _, doc := range docs {
+		if !opts.Deduplicater(ctx, doc) {
+			filtered = append(filtered, doc)
+		}
+	}
+	return filtered
+}
+
+// TODO :: Modify queries!
+func (vs *vectorStore) ApplyVectorIndex(ctx context.Context) error {
+	query := fmt.Sprintf(`
+        CREATE INDEX IF NOT EXISTS <changeTableName> ON <changeTableName>;`)
```

Comment:
I'm somewhat confused by the API of this function -- shouldn't it take in parameters for an index in order to work?